src/main/java/com/amazon/opendistroforelasticsearch/security/ssl/transport/OpenDistroPortUnificationHandler.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+package com.amazon.opendistroforelasticsearch.security.ssl.transport;
+
+
+import com.amazon.opendistroforelasticsearch.security.ssl.OpenDistroSecurityKeyStore;
+import com.amazon.opendistroforelasticsearch.security.ssl.util.SSLConnectionTestUtil;
+import com.amazon.opendistroforelasticsearch.security.ssl.util.SSLUtil;
+import com.google.common.annotations.VisibleForTesting;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.ChannelFutureListener;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPipeline;
+import io.netty.handler.codec.ByteToMessageDecoder;
+import io.netty.handler.ssl.SslHandler;
+import java.nio.charset.StandardCharsets;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import javax.net.ssl.SSLException;
+import java.util.List;
+
+/**
+ * Modifies the current pipeline dynamically to enable TLS
+ */
+public class OpenDistroPortUnificationHandler extends ByteToMessageDecoder {
```

Comment:
`OpenDistroPortUnificationHandler` -> `DualModeSSLHandler`, `ByteToMessageDecoder` -> `OptionalSslHandler`.

Reply:
We did consider using OptionalSslHandler, but that uses very weaker check to detect SSL i.e [getEncryptedPacketLength](https://github.com/netty/netty/blob/bfbeb2dec62102836e096e89cedf7b230517d786/handler/src/main/java/io/netty/handler/ssl/SslUtils.java#L185) method. where it checks for SSV2/tls 1.1 etc also (which are deprecated) which does not have proper header validation and does check value from predefined bytes to detect whether it is ssvl2 or not and it consider few non encrypted ES requests as SSLV2 messages.

Reply:
renaming OpenDistroPortUnificationHandler -> DualModeSSLHandler
however we cant use OptionalSslHandler instead of ByteToMessageDecoder since we are having our own implementation of the decode function in ByteToMessageDecoder in DualModeSSLHandler

Reply:
ODFE supports not only TLSv1.3, it is necessary to support other versions that openssl and Java supports.

Reply:
My bad, updated the comment, we ignore only SSLV2 and it is not technically a TLS protocol, from SSLv3 onwards it is made as TLS, to be backward compatible TLS version header continued to use major version of SSL i.e 3 and minor version to specify the version of TLS. 

major+minor version of TLS versions:

TLS1_VERSION           x'0301' # major version 3, minor version 1
TLS1_1_VERSION         x'0302' # major version 3, minor version 2
TLS1_2_VERSION         x'0303' # major version 3, minor version 3

Check that we currently do: https://github.com/vengadanathan-s/opendistro-security/blob/master/src/main/java/com/amazon/opendistroforelasticsearch/security/ssl/util/SSLUtil.java#L51-L66 , where we major version is 3 i.e TLS message or not. AFAIK It should not true for all TLS versions, but it wont detect SSLv2 (deprecated long and no longer used anywhere) .